nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java

```diff
@@ -0,0 +1,95 @@
+   * variables that have a {@code @Nullable} upper bound.
+   *
+   * @param tree the tree representing the instantiated type
+   * @param state visitor state
+   * @param analysis the analysis object
+   * @param config the analysis config
+   */
+  public static void checkInstantiationForParameterizedTypedTree(
+      ParameterizedTypeTree tree, VisitorState state, NullAway analysis, Config config) {
+    if (!config.isJSpecifyMode()) {
+      return;
+    }
+    List<? extends Tree> typeArguments = tree.getTypeArguments();
+    if (typeArguments.size() == 0) {
+      return;
+    }
+    Map<Integer, Tree> nullableTypeArguments = new HashMap<>();
+    for (int i = 0; i < typeArguments.size(); i++) {
+      Tree curTypeArg = typeArguments.get(i);
+      if (curTypeArg instanceof AnnotatedTypeTree) {
+        AnnotatedTypeTree annotatedType = (AnnotatedTypeTree) curTypeArg;
+        for (AnnotationTree annotation : annotatedType.getAnnotations()) {
+          Type annotationType = ASTHelpers.getType(annotation);
+          if (annotationType != null
+              && Nullness.isNullableAnnotation(annotationType.toString(), config)) {
+            nullableTypeArguments.put(i, curTypeArg);
+            break;
+          }
+        }
+      }
+    }
+    // base type that is being instantiated
+    Type baseType = ASTHelpers.getType(tree);
+    if (baseType == null) {
+      return;
+    }
+    com.sun.tools.javac.util.List<Type> baseTypeArgs = baseType.tsym.type.getTypeArguments();
+    for (int i = 0; i < baseTypeArgs.size(); i++) {
+      if (nullableTypeArguments.containsKey(i)) {
+
+        Type upperBound = baseTypeArgs.get(i).getUpperBound();
+        com.sun.tools.javac.util.List<Attribute.TypeCompound> annotationMirrors =
+            upperBound.getAnnotationMirrors();
+        boolean hasNullableAnnotation =
+            Nullness.hasNullableAnnotation(annotationMirrors.stream(), config);
+        // if base type argument does not have @Nullable annotation then the instantiation is
+        // invalid
+        if (!hasNullableAnnotation) {
+          invalidInstantiationError(nullableTypeArguments.get(i), state, analysis);
+        }
+      }
+    }
+    // recursive check for nested parameterized types
+    for (int i = 0; i < typeArguments.size(); i++) {
+      if (typeArguments.get(i) instanceof ParameterizedTypeTree) {
+        checkInstantiationForParameterizedTypedTree(
+            (ParameterizedTypeTree) typeArguments.get(i), state, analysis, config);
+      }
+    }
+  }
+
+  private static void invalidInstantiationError(Tree tree, VisitorState state, NullAway analysis) {
+    ErrorBuilder errorBuilder = analysis.getErrorBuilder();
+    ErrorMessage errorMessage =
+        new ErrorMessage(
+            ErrorMessage.MessageTypes.TYPE_PARAMETER_CANNOT_BE_NULLABLE,
+            "Generic type parameter cannot be @Nullable, as type variable does not have a @Nullable upper bound");
```

Comment:
Wonder if this message is informative enough, since it makes no reference to the actual code and which parameter cannot be `@Nullable`. Shouldn't this be something like:

> "Generic type parameter `@Nullable String` (pos: 0) in `MixedTypeParam<@Nullable String, String, String, String>` cannot be `@Nullable`, as type variable `E1` in `MixedTypeParam<E1, E2 extends @Nullable Object, E3 extends @Nullable Object, E4>` does not have a `@Nullable` upper bound."

Reply:
So we have designed the code so that the error location is exactly the generic type parameter with the `@Nullable` annotation; I think that should be enough to point out which type parameter is causing a problem?  We could try to give more info on the base type and type variable name in the error message.  Let me look into that.

Reply:
Ah, I hadn't realized that. Might be good to first see what the error message looks like currently on actual code. My main worry is that developers might miss which type parameter is the issue in a case where there are multiple, so if this points to the exact type parameter and that's always 100% clear to the developer seeing the error, that might be enough.

Reply:
I tweaked the error message in 3fb68563b2bc6a14e915df0cd505bd275a824d9c.  Now for [this line](https://github.com/uber/NullAway/blob/3fb68563b2bc6a14e915df0cd505bd275a824d9c/nullaway/src/test/java/com/uber/nullaway/NullAwayJSpecifyGenericsTests.java#L72) the following errors are printed:
```
/Test.java:8: warning: [NullAway] Generic type parameter cannot be @Nullable, as type variable E1 of type com.uber.Test.MixedTypeParam does not have a @Nullable upper bound
     static class PartiallyInvalidSubclass extends MixedTypeParam<@Nullable String, String, String, @Nullable String> {}
                                                                  ^
    (see http://t.uber.com/nullaway )

/Test.java:8: warning: [NullAway] Generic type parameter cannot be @Nullable, as type variable E4 of type com.uber.Test.MixedTypeParam does not have a @Nullable upper bound
     static class PartiallyInvalidSubclass extends MixedTypeParam<@Nullable String, String, String, @Nullable String> {}
```
WDYT?